File: src/openjarvis/traces/store.py
"""SQLite-backed trace storage."""

from __future__ import annotations

import json
import sqlite3
from pathlib import Path
from typing import Any, List, Optional

from openjarvis.core.events import Event, EventBus, EventType
from openjarvis.core.types import StepType, Trace, TraceStep

_CREATE_TRACES = """\
CREATE TABLE IF NOT EXISTS traces (
    id                   INTEGER PRIMARY KEY AUTOINCREMENT,
    trace_id             TEXT    NOT NULL UNIQUE,
    query                TEXT    NOT NULL DEFAULT '',
    agent                TEXT    NOT NULL DEFAULT '',
    model                TEXT    NOT NULL DEFAULT '',
    engine               TEXT    NOT NULL DEFAULT '',
    result               TEXT    NOT NULL DEFAULT '',
    outcome              TEXT,
    feedback             REAL,
    started_at           REAL    NOT NULL DEFAULT 0.0,
    ended_at             REAL    NOT NULL DEFAULT 0.0,
    total_tokens         INTEGER NOT NULL DEFAULT 0,
    total_latency_seconds REAL   NOT NULL DEFAULT 0.0,
    metadata             TEXT    NOT NULL DEFAULT '{}'
);
"""

_CREATE_STEPS = """\
CREATE TABLE IF NOT EXISTS trace_steps (
    id               INTEGER PRIMARY KEY AUTOINCREMENT,
    trace_id         TEXT    NOT NULL,
    step_index       INTEGER NOT NULL,
    step_type        TEXT    NOT NULL,
    timestamp        REAL    NOT NULL DEFAULT 0.0,
    duration_seconds REAL    NOT NULL DEFAULT 0.0,
    input            TEXT    NOT NULL DEFAULT '{}',
    output           TEXT    NOT NULL DEFAULT '{}',
    metadata         TEXT    NOT NULL DEFAULT '{}',
    FOREIGN KEY (trace_id) REFERENCES traces(trace_id)
);
"""

_CREATE_FTS = """\
CREATE VIRTUAL TABLE IF NOT EXISTS traces_fts USING fts5(
    trace_id, query, result, agent,
    content='traces',
    content_rowid='rowid',
    tokenize='unicode61'
);
"""

_FTS_SYNC_INSERT = """\
CREATE TRIGGER IF NOT EXISTS traces_fts_ai AFTER INSERT ON traces BEGIN
    INSERT INTO traces_fts(rowid, trace_id, query, result, agent)
    VALUES (new.rowid, new.trace_id, new.query, new.result, new.agent);
END;
"""

_INSERT_TRACE = """\
INSERT INTO traces (
    trace_id, query, agent, model, engine, result,
    outcome, feedback, started_at, ended_at,
    total_tokens, total_latency_seconds, metadata
) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
"""

_INSERT_STEP = """\
INSERT INTO trace_steps (
    trace_id, step_index, step_type, timestamp,
    duration_seconds, input, output, metadata
) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
"""


class TraceStore:
    """Append-only SQLite store for interaction traces."""

    def __init__(self, db_path: str | Path) -> None:
        self._db_path = str(db_path)
        self._conn = sqlite3.connect(self._db_path)
        self._conn.execute("PRAGMA journal_mode=WAL")
        self._conn.execute(_CREATE_TRACES)
        self._conn.execute(_CREATE_STEPS)
        self._conn.execute(_CREATE_FTS)
        self._conn.execute(_FTS_SYNC_INSERT)
        self._conn.commit()

    def save(self, trace: Trace) -> None:
        """Persist a complete trace with all its steps."""
        self._conn.execute(
            _INSERT_TRACE,
            (
                trace.trace_id,
                trace.query,
                trace.agent,
                trace.model,
                trace.engine,
                trace.result,
                trace.outcome,
                trace.feedback,
                trace.started_at,
                trace.ended_at,
                trace.total_tokens,
                trace.total_latency_seconds,
                json.dumps(trace.metadata),
            ),
        )
        for idx, step in enumerate(trace.steps):
            self._conn.execute(
                _INSERT_STEP,
                (
                    trace.trace_id,
                    idx,
                    step.step_type.value
                    if isinstance(step.step_type, StepType)
                    else step.step_type,
                    step.timestamp,
                    step.duration_seconds,
                    json.dumps(step.input),
                    json.dumps(step.output),
                    json.dumps(step.metadata),
                ),
            )
        self._conn.commit()

    def get(self, trace_id: str) -> Optional[Trace]:
        """Retrieve a trace by id, or ``None`` if not found."""
        row = self._conn.execute(
            "SELECT * FROM traces WHERE trace_id = ?", (trace_id,)
        ).fetchone()
        if row is None:
            return None
        return self._row_to_trace(row)

    def list_traces(
        self,
        *,
        agent: Optional[str] = None,
        model: Optional[str] = None,
        outcome: Optional[str] = None,
        since: Optional[float] = None,
        until: Optional[float] = None,
        limit: int = 100,
    ) -> List[Trace]:
        """Query traces with optional filters."""
        clauses: List[str] = []
        params: List[Any] = []
        if agent is not None:
            clauses.append("agent = ?")
            params.append(agent)
        if model is not None:
            clauses.append("model = ?")
            params.append(model)
        if outcome is not None:
            clauses.append("outcome = ?")
            params.append(outcome)
        if since is not None:
            clauses.append("started_at >= ?")
            params.append(since)
        if until is not None:
            clauses.append("started_at <= ?")
            params.append(until)
        where = " AND ".join(clauses) if clauses else "1=1"
        sql = f"SELECT * FROM traces WHERE {where} ORDER BY started_at DESC LIMIT ?"
        params.append(limit)
        rows = self._conn.execute(sql, params).fetchall()
        return [self._row_to_trace(r) for r in rows]

    def count(self) -> int:
        """Return the total number of stored traces."""
        row = self._conn.execute("SELECT COUNT(*) FROM traces").fetchone()
        return row[0] if row else 0

    def search(
        self,
        query: str,
        *,
        agent: str | None = None,
        limit: int = 20,
    ) -> list[dict[str, Any]]:
        """Full-text search across traces. Optionally filter by agent."""
        sql = (
            "SELECT t.trace_id, t.query, t.result, t.agent, t.model, t.outcome,"
            " t.started_at "
            "FROM traces_fts f JOIN traces t ON f.rowid = t.rowid "
            "WHERE traces_fts MATCH ?"
        )
        params: list[Any] = [query]
        if agent:
            sql += " AND t.agent = ?"
            params.append(agent)
        sql += " ORDER BY rank LIMIT ?"
        params.append(limit)
        rows = self._conn.execute(sql, params).fetchall()
        return [
            {
                "trace_id": r[0], "query": r[1], "result": r[2],
                "agent": r[3], "model": r[4], "outcome": r[5], "started_at": r[6],
            }
            for r in rows
        ]

    def subscribe_to_bus(self, bus: EventBus) -> None:
        """Subscribe to ``TRACE_COMPLETE`` events on *bus*."""
        bus.subscribe(EventType.TRACE_COMPLETE, self._on_event)

    def _on_event(self, event: Event) -> None:
        trace = event.data.get("trace")
        if isinstance(trace, Trace):
            self.save(trace)

    def update_feedback(self, trace_id: str, score: float) -> bool:
        """Update the feedback score for a trace.

        Returns True if the trace was found and updated, False otherwise.
        """
        cursor = self._conn.execute(
            "UPDATE traces SET feedback = ? WHERE trace_id = ?",
            (score, trace_id),
        )
        self._conn.commit()
        return cursor.rowcount > 0

    def close(self) -> None:
        """Close the underlying SQLite connection."""
        self._conn.close()

    # -- internal helpers ------------------------------------------------------

    def _row_to_trace(self, row: tuple) -> Trace:
        """Convert a traces table row + its steps into a Trace object."""
        trace_id = row[1]
        step_rows = self._conn.execute(
            "SELECT * FROM trace_steps WHERE trace_id = ? ORDER BY step_index",
            (trace_id,),
        ).fetchall()
        steps = [
            TraceStep(
                step_type=StepType(sr[3]),
                timestamp=sr[4],
                duration_seconds=sr[5],
                input=json.loads(sr[6]),
                output=json.loads(sr[7]),
                metadata=json.loads(sr[8]),
            )
            for sr in step_rows
        ]
        return Trace(
            trace_id=trace_id,
            query=row[2],
            agent=row[3],
            model=row[4],
            engine=row[5],
            result=row[6],
            outcome=row[7],
            feedback=row[8],
            started_at=row[9],
            ended_at=row[10],
            total_tokens=row[11],
            total_latency_seconds=row[12],
            metadata=json.loads(row[13]),
            steps=steps,
        )

    def _fetchall(self, sql: str = "SELECT * FROM traces") -> list:
        return self._conn.execute(sql).fetchall()


__all__ = ["TraceStore"]

File: repos/openjarvis/rust/crates/openjarvis-traces/src/store.rs
//! TraceStore — SQLite persistence for traces.

use openjarvis_core::{OpenJarvisError, Trace};
use parking_lot::Mutex;
use rusqlite::Connection;
use std::path::{Path, PathBuf};

pub struct TraceStore {
    conn: Mutex<Connection>,
    _db_path: PathBuf,
}

impl TraceStore {
    pub fn new(db_path: &Path) -> Result<Self, OpenJarvisError> {
        if let Some(parent) = db_path.parent() {
            std::fs::create_dir_all(parent).map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(e))
            })?;
        }

        let conn = Connection::open(db_path).map_err(|e| {
            OpenJarvisError::Io(std::io::Error::other(e.to_string()))
        })?;

        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS traces (
                trace_id TEXT PRIMARY KEY,
                query TEXT DEFAULT '',
                agent TEXT DEFAULT '',
                model TEXT DEFAULT '',
                engine TEXT DEFAULT '',
                result TEXT DEFAULT '',
                outcome TEXT,
                feedback REAL,
                started_at REAL DEFAULT 0,
                ended_at REAL DEFAULT 0,
                total_tokens INTEGER DEFAULT 0,
                total_latency_seconds REAL DEFAULT 0,
                steps_json TEXT DEFAULT '[]',
                metadata_json TEXT DEFAULT '{}'
            )",
        )
        .map_err(|e| {
            OpenJarvisError::Io(std::io::Error::other(e.to_string()))
        })?;

        Ok(Self {
            conn: Mutex::new(conn),
            _db_path: db_path.to_path_buf(),
        })
    }

    pub fn in_memory() -> Result<Self, OpenJarvisError> {
        Self::new(Path::new(":memory:"))
    }

    pub fn save(&self, trace: &Trace) -> Result<(), OpenJarvisError> {
        let steps_json = serde_json::to_string(&trace.steps).unwrap_or_default();
        let metadata_json = serde_json::to_string(&trace.metadata).unwrap_or_default();

        let conn = self.conn.lock();
        conn.execute(
            "INSERT OR REPLACE INTO traces
                (trace_id, query, agent, model, engine, result, outcome, feedback,
                 started_at, ended_at, total_tokens, total_latency_seconds,
                 steps_json, metadata_json)
            VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14)",
            rusqlite::params![
                trace.trace_id,
                trace.query,
                trace.agent,
                trace.model,
                trace.engine,
                trace.result,
                trace.outcome,
                trace.feedback,
                trace.started_at,
                trace.ended_at,
                trace.total_tokens,
                trace.total_latency_seconds,
                steps_json,
                metadata_json,
            ],
        )
        .map_err(|e| {
            OpenJarvisError::Io(std::io::Error::other(e.to_string()))
        })?;

        Ok(())
    }

    pub fn get(&self, trace_id: &str) -> Result<Option<Trace>, OpenJarvisError> {
        let conn = self.conn.lock();
        let mut stmt = conn
            .prepare(
                "SELECT trace_id, query, agent, model, engine, result, outcome, feedback,
                        started_at, ended_at, total_tokens, total_latency_seconds,
                        steps_json, metadata_json
                 FROM traces WHERE trace_id = ?1",
            )
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;

        let result = stmt
            .query_row(rusqlite::params![trace_id], |row| {
                Ok(Trace {
                    trace_id: row.get(0)?,
                    query: row.get(1)?,
                    agent: row.get(2)?,
                    model: row.get(3)?,
                    engine: row.get(4)?,
                    result: row.get(5)?,
                    outcome: row.get(6)?,
                    feedback: row.get(7)?,
                    started_at: row.get(8)?,
                    ended_at: row.get(9)?,
                    total_tokens: row.get(10)?,
                    total_latency_seconds: row.get(11)?,
                    steps: row
                        .get::<_, String>(12)
                        .ok()
                        .and_then(|s| serde_json::from_str(&s).ok())
                        .unwrap_or_default(),
                    metadata: row
                        .get::<_, String>(13)
                        .ok()
                        .and_then(|s| serde_json::from_str(&s).ok())
                        .unwrap_or_default(),
                })
            })
            .ok();

        Ok(result)
    }

    pub fn list_traces(
        &self,
        limit: usize,
        offset: usize,
    ) -> Result<Vec<Trace>, OpenJarvisError> {
        let conn = self.conn.lock();
        let mut stmt = conn
            .prepare(
                "SELECT trace_id, query, agent, model, engine, result, outcome, feedback,
                        started_at, ended_at, total_tokens, total_latency_seconds,
                        steps_json, metadata_json
                 FROM traces ORDER BY started_at DESC LIMIT ?1 OFFSET ?2",
            )
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;

        let traces = stmt
            .query_map(rusqlite::params![limit as i64, offset as i64], |row| {
                Ok(Trace {
                    trace_id: row.get(0)?,
                    query: row.get(1)?,
                    agent: row.get(2)?,
                    model: row.get(3)?,
                    engine: row.get(4)?,
                    result: row.get(5)?,
                    outcome: row.get(6)?,
                    feedback: row.get(7)?,
                    started_at: row.get(8)?,
                    ended_at: row.get(9)?,
                    total_tokens: row.get(10)?,
                    total_latency_seconds: row.get(11)?,
                    steps: row
                        .get::<_, String>(12)
                        .ok()
                        .and_then(|s| serde_json::from_str(&s).ok())
                        .unwrap_or_default(),
                    metadata: row
                        .get::<_, String>(13)
                        .ok()
                        .and_then(|s| serde_json::from_str(&s).ok())
                        .unwrap_or_default(),
                })
            })
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?
            .filter_map(|r| r.ok())
            .collect();

        Ok(traces)
    }

    pub fn count(&self) -> Result<usize, OpenJarvisError> {
        let conn = self.conn.lock();
        let count: i64 = conn
            .query_row("SELECT COUNT(*) FROM traces", [], |row| row.get(0))
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;
        Ok(count as usize)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_trace_store_save_and_get() {
        let store = TraceStore::in_memory().unwrap();
        let trace = Trace {
            trace_id: "test123".into(),
            query: "What is 2+2?".into(),
            agent: "simple".into(),
            model: "qwen3:8b".into(),
            result: "4".into(),
            ..Default::default()
        };

        store.save(&trace).unwrap();
        let retrieved = store.get("test123").unwrap().unwrap();
        assert_eq!(retrieved.query, "What is 2+2?");
        assert_eq!(retrieved.result, "4");
    }

    #[test]
    fn test_trace_store_list() {
        let store = TraceStore::in_memory().unwrap();
        for i in 0..5 {
            let trace = Trace {
                trace_id: format!("t{}", i),
                query: format!("query {}", i),
                ..Default::default()
            };
            store.save(&trace).unwrap();
        }
        assert_eq!(store.count().unwrap(), 5);
        let list = store.list_traces(3, 0).unwrap();
        assert_eq!(list.len(), 3);
    }
}
